### ISSUE TITLE
Linter: gosec, Rule: G103 - Audit the use of unsafe block. Should we enable it?

### ISSUE THREAD

**Issue by zak-pawel** [influxdata/telegraf#12891]
### Use Case

This issue starts discussion about enabling:
- linter: [gosec](https://github.com/securego/gosec) - Inspects source code for security problems by scanning the Go AST.
- rule: [G103 - Audit the use of unsafe block](https://github.com/securego/gosec#usage)

Rule is mapped to [CWE-242: Use of Inherently Dangerous Function](https://cwe.mitre.org/data/definitions/242.html).

### Expected behavior

Decision if rule should be enabled or not.

### Actual behavior

For this rule no findings were found in current code.

### Additional info

For this rule no additional configuration can be provided.

**Comment by powersj** [comment]
100% add however...

> For this rule no findings were found in current code.

How did it not find the imports in inputs.win_eventlog and inputs.win_perf_counters, parsers.influx, etc?

**Comment by zak-pawel** [comment]
> How did it not find the imports in inputs.win_eventlog and inputs.win_perf_counters, parsers.influx, etc?

@powersj For two reasons:

1. Our `golangci-lint` is run only for linux. So we don't see any issues related to code for windows or darwin only. Please, see: https://github.com/influxdata/telegraf/pull/10782
2. We don't have this section configured in `.golangci.yml`:
```
  # Independently of option `exclude` we use default exclude patterns,
  # it can be disabled by this option.
  # To list all excluded by default patterns execute `golangci-lint run --help`.
  # Default: true.
  exclude-use-default: false
```
So all these findings are by default excluded:
```
# EXC0001 errcheck: Almost all programs ignore errors on these functions and in most cases it's ok
- Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked

# EXC0002 golint: Annoying issue about not having a comment. The rare codebase has such comments
- (comment on exported (method|function|type|const)|should have( a package)? comment|comment should be of the form)

# EXC0003 golint: False positive when tests are defined in package 'test'
- func name will be used as test\.Test.* by other packages, and that stutters; consider calling this

# EXC0004 govet: Common false positives
- (possible misuse of unsafe.Pointer|should have signature)

# EXC0005 staticcheck: Developers tend to write in C-style with an explicit 'break' in a 'switch', so it's ok to ignore
- ineffective break statement. Did you mean to break out of the outer loop

# EXC0006 gosec: Too many false-positives on 'unsafe' usage
- Use of unsafe calls should be audited

# EXC0007 gosec: Too many false-positives for parametrized shell calls
- Subprocess launch(ed with variable|ing should be audited)

# EXC0008 gosec: Duplicated errcheck checks
- (G104|G307)

# EXC0009 gosec: Too many issues in popular repos
- (Expect directory permissions to be 0750 or less|Expect file permissions to be 0600 or less)

# EXC0010 gosec: False positive is triggered by 'src, err := ioutil.ReadFile(filename)'
- Potential file inclusion via variable

# EXC0011 stylecheck: Annoying issue about not having a comment. The rare codebase has such comments
- (comment on exported (method|function|type|const)|should have( a package)? comment|comment should be of the form)

# EXC0012 revive: Annoying issue about not having a comment. The rare codebase has such comments
- exported (.+) should have comment( \(or a comment on this block\))? or be unexported

# EXC0013 revive: Annoying issue about not having a comment. The rare codebase has such comments
- package comment should be of the form "(.+)...

# EXC0014 revive: Annoying issue about not having a comment. The rare codebase has such comments
- comment on exported (.+) should be of the form "(.+)..."

# EXC0015 revive: Annoying issue about not having a comment. The rare codebase has such comments
- should have a package comment
```

**Comment by zak-pawel** [comment]
With `exclude-use-default: false` set in `.golangci.yml` following findings were found for this rule (for linux):
```
plugins/inputs/vsphere/vsphere_test.go:226:5                                       	gosec  	G103: 	Use of unsafe calls should be audited
plugins/inputs/vsphere/vsphere_test.go:279:5                                       	gosec  	G103: 	Use of unsafe calls should be audited
plugins/inputs/vsphere/vsphere_test.go:403:5                                       	gosec  	G103: 	Use of unsafe calls should be audited
plugins/inputs/vsphere/vsphere_test.go:492:6                                       	gosec  	G103: 	Use of unsafe calls should be audited
plugins/outputs/application_insights/application_insights.go:287:6                 	gosec  	G103: 	Use of unsafe calls should be audited
plugins/outputs/postgresql/sqltemplate/template.go:329:22                          	gosec  	G103: 	Use of unsafe calls should be audited
plugins/parsers/influx/escape.go:85:33                                             	gosec  	G103: 	Use of unsafe calls should be audited
plugins/parsers/influx/escape.go:90:18                                             	gosec  	G103: 	Use of unsafe calls should be audited
```

**Comment by zak-pawel** [comment]
More info: https://securego.io/docs/rules/g103.html

**Comment by srebhan** [comment]
I also think we should enable this rule! Furthermore, can we run the linter on _all_ platforms like described [here](https://github.com/golangci/golangci-lint-action#multiple-os-support)?

**Comment by zak-pawel** [comment]
> Furthermore, can we run the linter on _all_ platforms like described [here](https://github.com/golangci/golangci-lint-action#multiple-os-support)?

@srebhan Yeah, we can. But first we need to:
- fix all findings for Windows (more than 500)
  - `win_perf_counters` is a nightmare
- probably rework few plugins to avoid "false-positive" (platform specific) `unused...` findings
  - look at findings for `diskio` end `ethtool` which are still there after https://github.com/influxdata/telegraf/pull/12958 (probably many more will be for Windows)

AFAIU, configuring linter for 3 platforms, will result in 3 different runs with 3 different result sets which will not be merged into one.

**Comment by zak-pawel** [comment]
> > Furthermore, can we run the linter on _all_ platforms like described [here](https://github.com/golangci/golangci-lint-action#multiple-os-support)?
> 
> @srebhan Yeah, we can. But first we need to:
> 
> * fix all findings for Windows (more than 500)
>   
>   * `win_perf_counters` is a nightmare
> * probably rework few plugins to avoid "false-positive" (platform specific) `unused...` findings
>   
>   * look at findings for `diskio` end `ethtool` which are still there after [chore: Fix linter findings for Darwin #12958](https://github.com/influxdata/telegraf/pull/12958) (probably many more will be for Windows)
> 
> AFAIU, configuring linter for 3 platforms, will result in 3 different runs with 3 different result sets which will not be merged into one.

@srebhan Regarding `golangci-lint` for Darwin and Windows I've created these: https://github.com/influxdata/telegraf/issues/13036 https://github.com/influxdata/telegraf/pull/13037
